compiler/rustc_type_ir/src/visit.rs

```diff
@@ -47,6 +47,83 @@ use std::ops::ControlFlow;
 
 use crate::{self as ty, BoundVars, Interner, IntoKind, Lrc, TypeFlags};
 
+/// Similar to the `Try` trait, but also implemented for `()`.
```

Comment:
Move this to rustc_data_structures, so you can avoid all the new crate dependencies (even if you have to add some, they'll already be there transitively)

Reply:
`rustc_type_ir` is already a transitive dependency for all of them via `rustc_ast` (or `rustc_hir`, which depends on `rustc_ast`).

Reply:
huh. Why does ast depend on type_ir

Reply:
Ah https://github.com/rust-lang/rust/blob/e9f95949138125bb1b98ed213a41b9aa825bacf5/compiler/rustc_ast/Cargo.toml#L17

Reply:
Let's avoid adding more direct deps and instead move this new trait and macros to data structures. It's not type system specific after all

Reply:
What's with the `nightly` feature on `rustc_type_ir` and all the optional dependencies? Should I just ignore those or make the whole visit module depend on a feature?

Reply:
Hmm I see your point. That's annoying. I guess it's the same with the movability and the mutability enums. If you feel really motivated rn, you could split those out into a new crate and get rid of the ast->type_ir dep

Reply:
Just pulling them into their own crate is easy. Doing it in a way that's actually nice is probably going to be a fair bit of work.

Reply:
The split is done. Please move it to rustc_ast_ir and adjust the dependencies. Where useful, reexports could help reduce the amount of direct dependencies on rustc_ast_ir